### PR TITLE
tests(core): verify that queues work after time adjustments

### PR DESCRIPTION
**Pull request by hanshuebner**
### Summary

Add a couple of tests that check whether queues still work when the system time is being changed forwards and backwards.

### Checklist

- [X] The Pull Request has tests
- [ ] There's an entry in the CHANGELOG (N/A tests only)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com (N/A tests only)

### CODE REVIEW

File: spec/01-unit/27-queue_spec.lua [hunk 1]
```diff
@@ -44,23 +42,49 @@ describe("plugin queue", function()
...
+          workspace = nil
+        },
+        -- We want to be able to fake the time returned by ngx.now() only in the queue module and leave everything
+        -- else alone so that we can see what effects changing the system time has on queues.
+        now = function()
+          local called_from = debug.getinfo(2, "nSl")
+          if string.match(called_from.short_src, "/queue.lua$") then
+            return real_now() + now_offset
+          else
+            return real_now()
```
Comment (jschmid1) [inline spec/01-unit/27-queue_spec.lua:81]: This _appears_ to be unnecessary. Monkeypatching a function should only affect the current scope of the test, let alone of the system.
If this way of isolating a module from getting patched, we should find a more generic way to alter system internals.

File: spec/01-unit/27-queue_spec.lua [hunk 2]
```diff
@@ -392,6 +416,128 @@ describe("plugin queue", function()
...
+          max_coalescing_delay = 10,
+        }),
+        function(_, entries)
+          number_of_entries_processed = number_of_entries_processed + #entries
+          return true
+        end,
+        nil,
+        i
+      )
+      now_offset = now_offset + now_offset
```
Comment (jschmid1) [inline spec/01-unit/27-queue_spec.lua:437]: What's the logic behind this?

File: spec/01-unit/27-queue_spec.lua [hunk 3]
```diff
@@ -392,6 +416,128 @@ describe("plugin queue", function()
...
+    end
+    helpers.wait_until(
+      function ()
+        return number_of_entries_processed == number_of_entries
+      end,
+      10
+    )
+  end)
+
+  it("works when time is moved backward while items are on the queue and not yet processed", function()
```
Comment (jschmid1) [inline spec/01-unit/27-queue_spec.lua:475]: I'd appreciate a little explanation of what's happening here.

File: spec/01-unit/27-queue_spec.lua (hunk 1 above)
Comment (hanshuebner) [reply spec/01-unit/27-queue_spec.lua:81]: The problem is that `wait_until` and other testing utilities don't like the time being manipulated under them.  If there is a better way to manipulate the time only for the code under test, not the testing environment, I'm open to suggestions.

File: spec/01-unit/27-queue_spec.lua (hunk 2 above)
Comment (hanshuebner) [reply spec/01-unit/27-queue_spec.lua:437]: We want to test whether the queue still works if entries are added to the queue while the current system time is changed (forward in this case).  This is just a simple way to achieve that, but I'm open to suggestions what would be better as a test.

File: spec/01-unit/27-queue_spec.lua (hunk 3 above)
Comment (hanshuebner) [reply spec/01-unit/27-queue_spec.lua:480]: Comments added.  Also a couple of sleeps while sending to cause the process to yield.

File: spec/01-unit/27-queue_spec.lua (hunk 1 above)
Comment (jschmid1) [reply spec/01-unit/27-queue_spec.lua:81]: I don't have a good solution either. Lets keep this on our radar.

**Review by jschmid1 — APPROVED** [review]

**Pull request merged**